Libraries/LibWeb/CSS/StyleValues/ShorthandStyleValue.cpp

```diff
@@ -119,6 +119,60 @@ String ShorthandStyleValue::to_string() const
+            values.append("historical-forms"sv);
+        auto caps_or_null = longhand(PropertyID::FontVariantCaps)->to_font_variant_caps();
+        if (caps_or_null.has_value() && caps_or_null.value() != CSS::FontVariantCaps::Normal) {
+            values.append(CSS::to_string(caps_or_null.release_value()));
+        }
+        auto east_asian = longhand(PropertyID::FontVariantEastAsian)->to_font_variant_east_asian();
+        if (!east_asian.normal) {
+            if (east_asian.variant != Gfx::FontVariantEastAsian::Variant::Unset)
+                values.append(Gfx::font_variant_east_asian_to_string(east_asian));
+            if (east_asian.width != Gfx::FontVariantEastAsian::Width::Unset)
+                values.append(Gfx::font_variant_east_asian_to_string(east_asian));
+        }
+        auto ligatures = longhand(PropertyID::FontVariantLigatures)->to_font_variant_ligatures();
+        if (!ligatures.normal && !ligatures.none) {
+            if (ligatures.common != Gfx::FontVariantLigatures::Common::Unset)
+                values.append(Gfx::font_variant_ligatures_to_string(ligatures));
+            if (ligatures.discretionary != Gfx::FontVariantLigatures::Discretionary::Unset) {
+                values.append(Gfx::font_variant_ligatures_to_string(ligatures));
+            }
+            if (ligatures.historical != Gfx::FontVariantLigatures::Historical::Unset) {
+                values.append(Gfx::font_variant_ligatures_to_string(ligatures));
+            }
+            if (ligatures.contextual != Gfx::FontVariantLigatures::Contextual::Unset) {
+                values.append(Gfx::font_variant_ligatures_to_string(ligatures));
+            }
+        }
+        auto numeric = longhand(PropertyID::FontVariantNumeric)->to_font_variant_numeric();
+        if (!numeric.normal) {
+            if (numeric.ordinal)
+                values.append("ordinal"sv);
+            if (numeric.slashed_zero)
+                values.append("slashed-zero"sv);
+            if (numeric.figure != Gfx::FontVariantNumeric::Figure::Unset)
+                values.append(Gfx::font_variant_numeric_to_string(numeric));
+            if (numeric.spacing != Gfx::FontVariantNumeric::Spacing::Unset)
+                values.append(Gfx::font_variant_numeric_to_string(numeric));
+            if (numeric.fraction != Gfx::FontVariantNumeric::Fraction::Unset)
+                values.append(Gfx::font_variant_numeric_to_string(numeric));
+        }
+        auto position_or_null = longhand(PropertyID::FontVariantPosition)->to_font_variant_position();
+        if (position_or_null.has_value() && position_or_null.value() != CSS::FontVariantPosition::Normal) {
+            values.append(CSS::to_string(position_or_null.release_value()));
+        }
+        StringBuilder builder;
+        if (values.is_empty())
+            builder.append("normal"sv);
+        else
+            builder.join(' ', values);
+        return MUST(builder.to_string());
```

Comment:
The WPT test: http://wpt.live/css/css-fonts/parsing/font-variant-valid.html is crashing here because the string it tries to construct is not valid UTF-8. I've not investigated thoroughly, but it looks like the data pointed to by one of the `StringView`s in `values` may have been freed.

Reply:
@tcl3 thanks for the review! I was not aware of these parsing tests, I considered writing something similar but I didn't have time. I've fixed the crashes and made sure most of them passes. There are a few left that's not implemented and a few tricky ones that can probably wait after landing the initial PR.